### PR TITLE
Introduce `SkipLastWhile` extension

### LINKED ISSUE #1036: Remove tail items while condition is true
Consider adding an operator that remove items from the tail end of a sequence while a given predicate returns true. The proposed name and signature would be:

```c#
public static IEnumerable<T> SkipLastWhile<T>(this IEnumerable<T> source, Func<T, bool> predicate)
```

Example:

```c#
var xs = new[]
{
    1, 2,
    0, 0, // will be preserved (not in tail position)
    3, 4,
    0, 0, // will be removed
};

var result = xs.SkipLastWhile(x => x is 0);

Console.WriteLine(result.ToDelimitedString(", "));
```

Output:

    1, 2, 0, 0, 3, 4

### PR DESCRIPTION
**Pull request by ArmoryNode**
Adds `SkipLastWhile` extension method and associated tests

Closes #1036

### CODE REVIEW

**Review by atifaziz — COMMENTED** [review]
This is a really good quality PR. Thanks for picking up #1085 and taking the trouble to conform to the project's conventions.

I've left a few comments & suggestion from my initial review.

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (viceroypenguin) [reply MoreLinq.Test/SkipLastWhileTest.cs:101]: Isn't this and `SkipLastWhileThrowsArgumentNullExceptionWhenPredicateIsNull` already covered by `NullArgumentTest`? Both of these tests should be removed altogether.

Comment (atifaziz) [reply MoreLinq.Test/SkipLastWhileTest.cs:101]: > Both of these tests should be removed altogether.

True.

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (ArmoryNode) [reply MoreLinq.Test/SkipLastWhileTest.cs:80]: I don't believe it's a necessary promise. Removed 👍

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (ArmoryNode) [reply MoreLinq.Test/SkipLastWhileTest.cs:89]: Thank you!

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (ArmoryNode) [reply MoreLinq.Test/SkipLastWhileTest.cs:101]: I had not seen this yet, very fancy! Removed the tests🙂

File: MoreLinq/SkipLastWhile.cs
Comment (ArmoryNode) [reply MoreLinq/SkipLastWhile.cs:47]: My original implementation was actually done using a `Queue`, but running initial benchmarks it seemed like the queue resulted in a slightly slower speed and a larger allocation. After running more benchmarks with more cases (empty sequences, only the first thousand items returned, all items returned, etc.) it looks like you're right, the queue is the better choice. _(especially since it can avoid Gen1 allocations altogether in certain cases as shown below)_.

| Method                  | N     | Mean      | Error    | StdDev   | Gen0    | Gen1    | Allocated |
|-------------------------|-------|-----------|----------|----------|---------|---------|-----------|
| SkipLastWhileArray      | 1,000 | 39.09 μs  | 0.070 μs | 0.062 μs | 2.5024  | 0.2441  | 39.16 KB  |
| SkipLastWhileList       | 1,000 | 42.63 μs  | 0.115 μs | 0.108 μs | 2.5024  | 0.2441  | 39.2 KB   |
| SkipLastWhileQueue      | 1,000 | 67.99 μs  | 0.226 μs | 0.200 μs | 0.4883  | -       | 8.35 KB   |
| SkipLastWhileStack      | 1,000 | 67.31 μs  | 0.234 μs | 0.207 μs | 0.4883  | -       | 8.34 KB   |

[LINQPad script](https://share.linqpad.net/3nf87km5.linq)

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (atifaziz) [inline MoreLinq.Test/SkipLastWhileTest.cs:27]: Since all tests pertain to `SkipLastWhile` that's also reflected in the test class name, you can remove repeating it as the prefix to each test case name.

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (atifaziz) [inline MoreLinq.Test/SkipLastWhileTest.cs:26]: Add a test case to demonstrate that `SkipWhileLast` is lazy execution semantics, like so:

```suggestion
        [Test]
        public void IsLazy()
        {
            _ = new BreakingSequence<object>().SkipLastWhile(BreakingFunc.Of<object, bool>());
        }

        [Test]
```

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (atifaziz) [inline MoreLinq.Test/SkipLastWhileTest.cs:24]: Given that `SkipLastWhile` can be optimised for sequences, read-only lists and lists, you will want to test that the behaviour is identical irrespective of the source type. Have a look at some of the other tests to see how this is done, where we use [`TestExtensions.ToSourceKind`](https://github.com/morelinq/MoreLINQ/blob/03a160ceaa0942f46cf585894a3c6e53df040076/MoreLinq.Test/TestExtensions.cs#L87) to help with this.

**Review by atifaziz — COMMENTED** [review]
Thanks for sharing [your findings](https://github.com/morelinq/MoreLINQ/pull/1085#discussion_r1822982900).

I have done a second review pass and submitted some suggestions and comments.

I reckon we need one final pass and then we can get this in! 🏁

---
PS Run `build.cmd` or `build.sh` locally to update the generated extensions since the [CI checks/runs are failing due to stale code](https://github.com/morelinq/MoreLINQ/actions/runs/11597545671/job/32441333996#step:15:40).

File: MoreLinq/SkipLastWhile.cs
Comment (ArmoryNode) [reply MoreLinq/SkipLastWhile.cs:62]: I do like the idea of optimizing for specific types instead of treating them all the same, but I do have a question about `.TryAsListLike()`. Would checking for the count on line `44` potentially cause a multiple enumeration if the count is greater than 0 and `IterateList` is called? It's probably a non-issue, but I would like to know what your thoughts are on the potential performance implications🙂

Comment (atifaziz) [reply MoreLinq/SkipLastWhile.cs:62]: > Would checking for the count on line `44` potentially cause a multiple enumeration if the count is greater than 0 and `IterateList` is called?

No. The `Count` is checked without any iteration _if_ the runtime type supports querying the count:

https://github.com/morelinq/MoreLINQ/blob/03a160ceaa0942f46cf585894a3c6e53df040076/MoreLinq/ListLike.cs#L61-L68

`TestingSequence` also has a built-in check against multiple iterations:

https://github.com/morelinq/MoreLINQ/blob/03a160ceaa0942f46cf585894a3c6e53df040076/MoreLinq.Test/TestingSequence.cs#L92

`maxEnumerations` is set to a default of 1 in `TestingSequence.Of`:

https://github.com/morelinq/MoreLINQ/blob/03a160ceaa0942f46cf585894a3c6e53df040076/MoreLinq.Test/TestingSequence.cs#L29-L33

Hope this clears up the mystery.

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (ArmoryNode) [reply MoreLinq.Test/SkipLastWhileTest.cs:24]: Due to `BreakingCollection` and `BreakingReadOnlyCollection` not being list-like, the test cases using them fail since `IterateSequence` uses a foreach loop, since it will get the enumerator. Is it possible to successfully test for these cases?

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (atifaziz) [inline MoreLinq.Test/SkipLastWhileTest.cs:99]: Unless I've missed something, this seems redundant with `PredicateBecomesTruePartWay` and therefore can be completely removed:

```suggestion
```

There is nothing apparent here about the optimisation.

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (atifaziz) [inline MoreLinq.Test/SkipLastWhileTest.cs:1]: nit: There's a UTF-8 BOM marker at the beginning of this file [that shouldn't be there](https://github.com/morelinq/MoreLINQ/blob/1726b52594c93e342bb629fb7652d50426815851/.editorconfig#L23).

**Review by atifaziz — CHANGES_REQUESTED** [review]
Apart from the comments I've left, this needs merging with the base since [.NET 9 has been added as a new target](https://github.com/morelinq/MoreLINQ/pull/1091) and the [public API](https://github.com/morelinq/MoreLINQ/blob/1726b52594c93e342bb629fb7652d50426815851/MoreLinq/PublicAPI/net9.0/PublicAPI.Unshipped.txt) will need entries for this new extension. After that, I think we're good to merge. Thanks!

File: MoreLinq.Test/SkipLastWhileTest.cs
Comment (ArmoryNode) [reply MoreLinq.Test/SkipLastWhileTest.cs:24]: Got it. I appreciate your patience. 🙂

**Review by atifaziz — APPROVED** [review]
Thanks for your patience, all the follow-up and contributing this in the first place.

**Pull request merged**
